### PR TITLE
Add preliminary support for CodeGen models and start of fast inference

### PR DESCRIPTION
**Pull request by claudiosv**

### CODE REVIEW

**Review by claudiosv — COMMENTED** [review]
Please let me know if this is suitable to merge.

File: lmwrapper/huggingface_wrapper.py [hunk 1]
```diff
@@ -175,13 +302,250 @@ def _gather_logprobs_from_logits(
...
+
+
+def get_ort_model(model: PreTrainedModel) -> "ORTModel":
+    if model in {T5ForConditionalGeneration, AutoModelForSeq2SeqLM}:
+        return ORTModelForSeq2SeqLM
+
+    return ORTModelForCausalLM
+
+
+def initialize_hf_model(
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:402]: Should this be private? Else maybe we need to name or document a bit a clearer how the difference between these two methods for the end user of this module. I can see usecases when a user might want more control over initialization, but since we are still making progress on this part, private might make the most sense for now until we get a more stable API.

File: lmwrapper/huggingface_wrapper.py [hunk 2]
```diff
@@ -175,13 +302,250 @@ def _gather_logprobs_from_logits(
...
+    return predictor
+
+
+def warmup_model(predictor: HuggingfacePredictor):
+    small_prompt = LmPrompt("!", cache=False, temperature=0)
+    predictor.predict(small_prompt)
+
+    long_prompt_str = (
+        "hello world" * predictor.get_model_max_length()
+    )  # TODO: this should be valid model tokens!
```
Comment (DNGros) [inline lmwrapper/huggingface_wrapper.py:549]: I'd raise a NotImplemented error in this module. This will almost certainly fail with an inscrutable error message since the "hello world" likely becomes multiple tokens. Until we add this part probably best to guard with NotImplemented error rather than only the TODO comment in this context.

File: test/test_huggingface.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    InstructCodeT5plus_16B = "Salesforce/instructcodet5p-16b"
+
+    DistilGPT2 = "distilgpt2"
+    GPT2 = "gpt2"
+
+
+CUDA_UNAVAILABLE = not torch.cuda.is_available()
+SMALL_GPU = (
+    CUDA_UNAVAILABLE or torch.cuda.mem_get_info()[0] < 17_179_869_184
+)  # 16GB
```
Comment (DNGros) [inline test/test_huggingface.py:24]: Sorry to bring this back up, but I am not a big fan of the behaviour of tests depending on the runtime environment. When we read a test passes, we should know what that means. I think we should parametrize with truely all models. Then within a given test can call `pytest.skip()` within the method in cases when in the SMALL_GPU case. This means the test report will always more explicitly reveal what the test actually did and distinguish between all tests truely passing and not.

File: setup.py [hunk 4]
```diff
@@ -1,25 +1,38 @@
...
-    python_requires='>=3.10',
+    name="lmwrapper",
+    version="0.03.01",
+    author="David Gros",
+    description="Wrapper around language model APIs",
+    license="MIT",
+    packages=["lmwrapper"],
+    install_requires=get_requirments("requirements.txt"),
+    extras_require={
+        "huggingface": get_requirments("requirements-hf.txt"),
```
Comment (DNGros) [inline setup.py:25]: I expect we need to add the [huggingface] arg in the github workflow.

Also, please update the install instructions in the readme to note the option to install with huggingface support.

**Review by DNGros — COMMENTED** [review]
Cool looks good! Couple of small comments, but seems essentially ready to merge. Just make sure to test on our GPU server before merging since we aren't set up currently to test a lot of this stuff in CI.

File: lmwrapper/huggingface_wrapper.py (hunk 1 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:402]: Agreed.

File: lmwrapper/huggingface_wrapper.py (hunk 2 above)
Comment (claudiosv) [reply lmwrapper/huggingface_wrapper.py:549]: Agreed. Originally I had truncation enabled for the tokenizer but removed that since I agree we should not change defaults that cause unexpected behavior.

File: test/test_huggingface.py (hunk 3 above)
Comment (claudiosv) [reply test/test_huggingface.py:24]: Thanks for explaining. I think I get your point now and agree. Will fix.

File: lmwrapper/huggingface_wrapper.py [hunk 5]
```diff
@@ -1,24 +1,132 @@
...
+        # llm_int8_has_fp16_weight (bool, optional, defaults to False) — This flag runs LLM.int8() with 16-bit main weights. This is useful for fine-tuning as the weights do not have to be converted back and forth for the backward pass.
+        # bnb_4bit_compute_dtype (torch.dtype or str, optional, defaults to torch.float32) — This sets the computational type which might be different than the input time. For example, inputs might be fp32, but computation can be set to bf16 for speedups.
+        # bnb_4bit_quant_type (str, {fp4, nf4}, defaults to fp4) — This sets the quantization data type in the bnb.nn.Linear4Bit layers. Options are FP4 and NF4 data types which are specified by fp4 or nf4.
+        # bnb_4bit_use_double_quant (bool, optional, defaults to False) — This flag is used for nested quantization where the quantization constants from the first quantization are quantized again.
+        load_in_4bit=True,
+        bnb_4bit_use_double_quant=True,
+    )
+except ImportError:
+    print(
```
[the remark this answers is not in the document]
Comment (DNGros) [reply lmwrapper/huggingface_wrapper.py:55]: That can be fine. I don't have a lot of experience using python loggers (but probably should). I would say can either just use the standard builtin logger, or just `print(msg, file=sys.stderr)`. Even this feels a bit weird to warn as a there isn't any current support yet for quantization. This is a nit, so I wouldn't worry to much about this.

File: README.md [hunk 6]
```diff
@@ -8,8 +8,24 @@ you to switch inbetween openai endpoints and local models with minimal changes.
...
+```
+
+To take advantage of faster inference via [Optimum](https://github.com/huggingface/optimum) and [Xformers](https://github.com/facebookresearch/xformers):
+```bash
+pip install lmwrapper[ort] @ git+https://github.com/DNGros/lmwrapper.git
+```
+
+To take advantage of faster inference using the ONNX CUDA runtime on GPU(s):
+```bash
+pip install lmwrapper[ort-gpu] @ git+https://github.com/DNGros/lmwrapper.git
```
Comment (DNGros) [inline README.md:28]: We are not support these in this PR right? Why suggest it here?

Comment (claudiosv) [reply README.md:28]: Right, my bad.

**Comment by DNGros** [comment]
Ok cool. Looks like the test skipping behavior is my only lingering comment.
